contracts/Launchpad.sol

```diff
@@ -0,0 +1,492 @@
+        }
+    }
+
+    function setTiersRequirements(uint256[] calldata tierIds, uint256[] calldata stakingRequirements) external restricted {
+        require(tierIds.length <= vars[VAR_TIERS_AMOUNT], "Too many tiers");
+        require(tierIds.length == stakingRequirements.length, "Wrong input");
+        for (uint256 i = 0; i < tierIds.length; i++) {
+            require(tierIds[i] > 0 && tierIds[i] <= vars[VAR_TIERS_AMOUNT], "Wrong id");
+            tierStakingRequirement[tierIds[i]] = stakingRequirements[i];
+        }
+    }
+
+    function setTiersAllocationWeights(uint256[] calldata tierIds, uint256[] calldata tierWeights) external restricted {
+        require(tierIds.length <= vars[VAR_TIERS_AMOUNT], "Too many tiers");
+        require(tierIds.length == tierWeights.length, "Wrong input");
+        for (uint256 i = 0; i < tierIds.length; i++) {
+            require(tierIds[i] > 0 && tierIds[i] <= vars[VAR_TIERS_AMOUNT], "Wrong id");
+            tierAllocationWeight[tierIds[i]] = tierWeights[i];
+        }
+    }
+
+    // PROJECT DETAILS
+
+    function addNewLaunch(
+        string calldata name,
+        string calldata tokenSymbol,
+        string calldata detailsJsonUri,
+        string calldata imageUrl,
+        address fundingTokenAddress
+    ) external restricted {
+        launches[nextLaunchId] = Launch(
+            name,
+            tokenSymbol,
+            detailsJsonUri,
+            imageUrl,
+            fundingTokenAddress,
+            1
+        );
+
+        emit LaunchAdded(nextLaunchId, 1);
+        // nextLaunchId + 1 is reserved for Phase 2s
+        nextLaunchId += 2;
+    }
+
+    function addSecondPhaseForLaunch(uint256 launchId, uint256 startTime) external onlyPhase1(launchId) restricted {
+        Launch memory lp = launches[launchId];
+        require(launchId < nextLaunchId, "Wrong id");
+        require(block.timestamp > launchStartTime[launchId] + vars[VAR_FUNDING_PERIOD_PHASE_1], "Phase 1 not finished");
+        require(launchFundsToRaise[launchId].sub(launchTotalRaised[launchId]) > 0, "Tokens sold out");
+        launches[launchId + 1] = Launch(
+            lp.name,
+            lp.tokenSymbol,
+            lp.detailsJsonUri,
+            lp.imageUrl,
+            lp.fundingTokenAddress,
+            2
+        );
+        
+        launchStartTime[launchId + 1] = startTime;
+        launchTokenPrice[launchId + 1] = launchTokenPrice[launchId];
+        launchFundsToRaise[launchId + 1] = launchFundsToRaise[launchId].sub(launchTotalRaised[launchId]);
+        launchBaseAllocation[launchId + 1] = launchBaseAllocation[launchId];
+        launchTokenAddress[launchId + 1] = launchTokenAddress[launchId];
+        launchBaseAllocation[launchId + 1] = launchBaseAllocation[launchId];
+
+        emit LaunchAdded(launchId + 1, 2);
+    }
+
+    function removeLaunch(uint256 launchId) external restricted {
+        delete launches[launchId];
+    }
+
+    // UPDATING DETAILS
```

Comment:
There is no check that the launch id is valid (> 0 < next launch) but should be fine as these are dev only and data can be removed if set incorrectly

Reply:
I think it's fine, saves a bit on contract size to skip the check.